File: backend/internal/model/db/projection/prepaid.go
package projection

import (
	"akatengu/internal/enums"

	"github.com/shopspring/decimal"
)

//dbmap:sqlcdb=Prepaid
type Prepaid struct {
	ID               int64               `db:"id" json:"id"`
	MerchantID       int64               `db:"merchant_id" json:"merchant_id"`
	TxnID            *int64              `db:"txn_id" json:"txn_id"`
	AccountID        string              `db:"account_id" json:"account_id"`
	ExpenseAccountID string              `db:"expense_account_id" json:"expense_account_id"`
	Name             string              `db:"name" json:"name"`
	TotalAmount      decimal.Decimal     `db:"total_amount" json:"total_amount"`
	AmortizedAmount  decimal.Decimal     `db:"amortized_amount" json:"amortized_amount"`
	Periods          int64               `db:"periods" json:"periods"`
	AmortizedPeriods int64               `db:"amortized_periods" json:"amortized_periods"`
	StartDate        string              `db:"start_date" json:"start_date"`
	Status           enums.PrepaidStatus `db:"status" json:"status"`
	UpdatedBy        *string             `db:"updated_by" json:"updated_by"`
	UpdatedAt        *string             `db:"updated_at" json:"updated_at"`
	Version          int64               `db:"version" json:"version"`
}

//dbmap:sqlcdb=PrepaidAmortization
type PrepaidAmortization struct {
	ID         int64           `db:"id" json:"id"`
	MerchantID int64           `db:"merchant_id" json:"merchant_id"`
	PrepaidID  int64           `db:"prepaid_id" json:"prepaid_id"`
	TxnID      int64           `db:"txn_id" json:"txn_id"`
	PeriodDate string          `db:"period_date" json:"period_date"`
	Amount     decimal.Decimal `db:"amount" json:"amount"`
	UpdatedBy  *string         `db:"updated_by" json:"updated_by"`
	UpdatedAt  *string         `db:"updated_at" json:"updated_at"`
}

File: frontside/src/types/prepaid.ts
export type PrepaidStatus = 'ACTIVE' | 'COMPLETED' | 'DISPOSED';

export interface Prepaid {
  id:                 number;
  prepaid_uuid:       string;
  merchant_id:        number;
  txn_id:             number | null;
  account_id:         string;
  expense_account_id: string;
  name:               string;
  total_amount:       string;
  amortized_amount:   string;
  periods:            number;
  amortized_periods:  number;
  start_date:         string;
  status:             PrepaidStatus;
  updated_by:         string | null;
  updated_at:         string | null;
  version:            number;
}

export interface PrepaidAmortization {
  id:          number;
  merchant_id: number;
  prepaid_id:  number;
  txn_id:      number;
  period_date: string;
  amount:      string;
  updated_by:  string | null;
  updated_at:  string | null;
}

export interface PrepaidCreatedPayload {
  category_uuid: string;
  ledger_uuid:   string;
  name:          string;
  total_amount:  string;
  periods:       number;
  start_date:    string;
  memo:          string;
  note:          string;
}

import type { InstallmentTermsPayload } from './installment';

export interface PrepaidCreatedWithInstallmentPayload {
  category_uuid: string;
  name:          string;
  total_amount:  string;
  periods:       number;
  start_date:    string;
  memo:          string;
  note:          string;
  installment:   InstallmentTermsPayload;
}

export interface PrepaidAmortizedPayload {
  prepaid_uuid: string;
  period_date:  string;
}

export interface PrepaidDisposedPayload {
  prepaid_uuid:  string;
  disposal_date: string;
  memo:          string;
}
